sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1128,6 +1131,18 @@ static int create_sdw_dailink(struct snd_soc_card *card,
 			cpus[cpu_dai_index++].dai_name = cpu_name;
 		}
 
+		for (index = 0; index < *link_index; index++) {
+			/*
+			 * Append codec type to dai link name if the dai link name is duplicated.
+			 */
+			if (!strcmp(name, dai_links[index].name)) {
+				devm_kfree(dev, name);
+				name = devm_kasprintf(dev, GFP_KERNEL, "%s-%s",
+						      dai_links[index].name,
+						      type_strings[codec_info_list[codec_index].codec_type]);
```

Comment:
> @ujfalusi for the current case @bardliao is trying to solve, the dmic and jack are independent SoundWire codecs that happen to be in the same package. So there should be no confusion here.
> 
> However when we have to support SDCA, we will need to have multiple functions, so we should probably add a hierarchy such as
> 
> ```
> dailink
>     codec[k]
>         function[l]
>             dai[m]
> ```
> 
> with ALC712, the 'k' value would be 2, 'l' would be 1 and 'm' is 1 as well.

@plbossart I have a proposal on https://github.com/thesofproject/linux/pull/4191. The idea is to let dai_name be an array in sof_sdw_codec_info {} and create a dai link for each dai.

Reply:
> commit tile: "ASoC: Intel: sof_sdw: ..."
> 
> but this will be confusing, right? Let's say we have the two DAI on codec0:
> 
> 1. jack
> 2. mic
> 
> one of the name will be "codec0" the other will be "codec0-mic" or "codec0-jack" depending on the order they are added, so one should be guessing what the "codec0" is?

@ujfalusi IMHO. we don't rely on the "type" to know the dai link function. There is no "type" in the existing dai link name anyway.  In most cases, the dai link id will be assigned based on its function, that is
```
0: jack out
1: jack in
2: amp
3: amp feedback
4: mic
```

Reply:
> @plbossart I have a proposal on #4191. The idea is to let dai_name be an array in sof_sdw_codec_info {} and create a dai link for each dai.

@bardliao I think we could improve a bit with more hierarchy, something like

`
.part_id = 0xdead,
.dai_num = 2,
.dai_descriptor = {
   { 
      {
    	.direction = {true, true},
        .dai_name = "jack-aif1",
	.init = sof_sdw_jack_init,
        .dai_type = SOF_SDW_CODEC_TYPE_JACK,
        .dailink = {0 , 1},
    },
    {
    	.direction = {true, true},
        .dai_name = "amp-aif1",
	.init = sof_sdw_amp_init,
        .dai_type = SOF_SDW_CODEC_TYPE_AMP,
        .dalink = {2, 3},
    },
}
`

Reply:
> > @plbossart I have a proposal on #4191. The idea is to let dai_name be an array in sof_sdw_codec_info {} and create a dai link for each dai.
> 
> @bardliao I think we could improve a bit with more hierarchy, something like
> 
> ```
> .part_id = 0xdead,
> .dai_num = 2,
> .dai_descriptor = {
>    { 
>     	.direction = {true, true},
>         .dai_name = "jack-aif1",
> 	.init = sof_sdw_jack_init,
>         .dai_type = SOF_SDW_CODEC_TYPE_JACK,
>         .dailink = {0 , 1},
>     },
>     {
>     	.direction = {true, true},
>         .dai_name = "amp-aif1",
> 	.init = sof_sdw_amp_init,
>         .dai_type = SOF_SDW_CODEC_TYPE_AMP,
>         .dalink = {2, 3},
>     },
> }
> ```

@plbossart .dailink = {0 , 1}, means dai link id = 0 and 1, right? I am wondering do we need to specify the dai link id here.  We arrange dai link id by type. I think it is enough, right?
Another open is that rt712 supports jack and amp function, but customer may use rt712's jack function + another amp like rt1308. How should we handle this configuration? Currently, my idea is to create rt712 amp dai link anyway and topology can ignore the dai link. What do you think?

Reply:
Good question @bardliao. My point was that we keep struggling to maintain constant dailink values, depending on configurations we have to change the code to keep the indices as what the topology expects them.

So I was wondering if there's any merit in adding the expected dailink number directly in the descriptor. It's a suggestion only.

I am not sure how we would deal with a case with two amplifier dailinks, one for the 712 amp (not used in practice) and one for the rt1308. There's clearly a conflict with dailink numbers, the topology knows nothing about the codec so which of the two would be selected?

I also wonder if this topology makes any sense. rt712 is mainly for cost-savings with a single-chip solution, OEMs wouldn't add an additional amplifier and not use the RT712 amplifier. What's possible though is that additional amplifiers are used for secondary transducers, but in that case they would be on the same dailink. We probably want to deal with simple configurations first....

Reply:
> Good question @bardliao. My point was that we keep struggling to maintain constant dailink values, depending on configurations we have to change the code to keep the indices as what the topology expects them.
> 
> So I was wondering if there's any merit in adding the expected dailink number directly in the descriptor. It's a suggestion only.

Having dailink id in codec_info_list[] is indeed a good idea. My only concern is that someone might design a device with two jack/amp/mic. But it is not likely happened in real world.

> 
> I am not sure how we would deal with a case with two amplifier dailinks, one for the 712 amp (not used in practice) and one for the rt1308. There's clearly a conflict with dailink numbers, the topology knows nothing about the codec so which of the two would be selected?

The amps should be aggregated even they are not the same module, right? 

> 
> I also wonder if this topology makes any sense. rt712 is mainly for cost-savings with a single-chip solution, OEMs wouldn't add an additional amplifier and not use the RT712 amplifier. What's possible though is that additional amplifiers are used for secondary transducers, but in that case they would be on the same dailink. We probably want to deal with simple configurations first....

Agree, let's focus on the known configuration first. I think this PR is good enough to support rt713.